src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-object.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Management.Automation;
+using System.Management.Automation.Internal;
+using System.Management.Automation.Language;
+using System.Text;
+
+namespace Microsoft.PowerShell.Commands.Utility
+{
+    /// <summary>
+    /// Join-Object implementation.
+    /// </summary>
+    [Cmdlet(VerbsCommon.Join, "Object", RemotingCapability = RemotingCapability.None, DefaultParameterSetName = "default")]
+    [OutputType(typeof(string))]
+    public sealed class JoinObjectCommand : PSCmdlet
+    {
+        private const int Capacity = 50;
+
+        // ReSharper disable once CollectionNeverQueried.Local
+        private readonly List<PSObject> _inputObjects = new List<PSObject>(Capacity);
+        private DynamicPropertyGetter _propGetter = new DynamicPropertyGetter();
+
+        /// <summary>
+        /// Gets or sets the property name or script block to use as the value to join.
+        /// </summary>
+        [Parameter(Position = 0)]
+        [ArgumentCompleter(typeof(PropertyNameCompleter))]
+        public object PropertyName { get; set; }
```

Comment:
Can we use PSPropertyExpression type like in #6934?

Reply:
PropertyExpression has some extra semantics, handling hashtables like `@{n="x";ex={2,3,4}}`.

I looked at it, but was not convinced it was a fit. But there's definitely an overlap. Input here is most welcome.